### PR TITLE
relates to #1730: document write resource tests

### PR DESCRIPTION
**Pull request by EricBorczuk**
**What this PR does**:

**Which issue(s) this PR fixes**:
Fixes #<issue number>

**Checklist**
- [ ] Changes manually tested
- [ ] Automated Tests added/updated
- [ ] Documentation added/updated
- [ ] CLA Signed: [DataStax CLA](https://cla.datastax.com/)

### CODE REVIEW

**Comment by ivansenic** [comment]
@EricBorczuk I am re-basing this in order to trigger the new CI..

File: sgv2-docsapi/src/test/java/io/stargate/sgv2/docsapi/api/v2/namespaces/collections/documents/DocumentWriteResourceIntegrationTest.java [hunk 1]
```diff
@@ -0,0 +1,388 @@
...
+          .body(DEFAULT_PAYLOAD)
+          .when()
+          .post(BASE_PATH, DEFAULT_NAMESPACE, DEFAULT_COLLECTION)
+          .then()
+          .statusCode(201)
+          .body("profile", notNullValue());
+    }
+
+    @Test
+    public void withTtl() {
```
Comment (ivansenic) [inline sgv2-docsapi/src/test/java/io/stargate/sgv2/docsapi/api/v2/namespaces/collections/documents/DocumentWriteResourceIntegrationTest.java:115]: should we also test for longer ttls that will not delete the document just after write

File: sgv2-docsapi/src/test/java/io/stargate/sgv2/docsapi/api/v2/namespaces/collections/documents/DocumentWriteResourceIntegrationTest.java [hunk 2]
```diff
@@ -0,0 +1,388 @@
...
+        .atMost(Duration.ofSeconds(10));
+
+    tableManager
+        .createCollectionTable(DEFAULT_NAMESPACE, DEFAULT_COLLECTION)
+        .await()
+        .atMost(Duration.ofSeconds(10));
+  }
+
+  @Nested
+  class WriteDocument {
```
Comment (ivansenic) [inline sgv2-docsapi/src/test/java/io/stargate/sgv2/docsapi/api/v2/namespaces/collections/documents/DocumentWriteResourceIntegrationTest.java:74]: missing cases:

* write to a collection that does not exist
* write empty object
* write non-empty array
* write empty array
* write primitive JSON
* write without body

File: sgv2-docsapi/src/test/java/io/stargate/sgv2/docsapi/api/v2/namespaces/collections/documents/DocumentWriteResourceIntegrationTest.java [hunk 3]
```diff
@@ -0,0 +1,388 @@
...
+              .when()
+              .post(BASE_PATH, DEFAULT_NAMESPACE, DEFAULT_COLLECTION)
+              .peek();
+
+      assertThat(postResponse.statusCode()).isEqualTo(201);
+      assertThat(postResponse.header("location")).isNotNull();
+      String location = postResponse.header("location");
+
+      Awaitility.await()
+          .atLeast(0, TimeUnit.MILLISECONDS)
```
Comment (ivansenic) [inline sgv2-docsapi/src/test/java/io/stargate/sgv2/docsapi/api/v2/namespaces/collections/documents/DocumentWriteResourceIntegrationTest.java:131]: `atLeast` is not used correct here (https://github.com/awaitility/awaitility/wiki/Usage#at-least), actually has no affect.. Remove.. 

`atMost` with 2 seconds (1s ttl +1s buffer) seems like the correct way imo

File: sgv2-docsapi/src/test/java/io/stargate/sgv2/docsapi/api/v2/namespaces/collections/documents/DocumentWriteResourceIntegrationTest.java [hunk 4]
```diff
@@ -0,0 +1,388 @@
...
+              objectMapper.readTree(batchResponse.body().asString()).requiredAt("/documentIds");
+      Set<String> expectedIds = new HashSet<>();
+      expectedIds.add("1");
+      expectedIds.add("2");
+      expectedIds.add("3");
+      for (JsonNode node : documentIds) {
+        String id = node.asText();
+        expectedIds.remove(id);
+      }
+      assertThat(expectedIds).isEmpty();
```
Comment (ivansenic) [inline sgv2-docsapi/src/test/java/io/stargate/sgv2/docsapi/api/v2/namespaces/collections/documents/DocumentWriteResourceIntegrationTest.java:229]: all of this with:

```
.then()
.statusCode(202)
.body("documentIds", contains("1", "2", "3"))
```

File: sgv2-docsapi/src/test/java/io/stargate/sgv2/docsapi/api/v2/namespaces/collections/documents/DocumentWriteResourceIntegrationTest.java [hunk 5]
```diff
@@ -0,0 +1,388 @@
...
+              .peek();
+
+      assertThat(postResponse.statusCode()).isEqualTo(202);
+      assertThat(postResponse.body()).isNotNull();
+      ArrayNode ids =
+          (ArrayNode)
+              objectMapper.readTree(postResponse.body().asString()).requiredAt("/documentIds");
+
+      Awaitility.await()
+          .atLeast(0, TimeUnit.MILLISECONDS)
```
Comment (ivansenic) [inline sgv2-docsapi/src/test/java/io/stargate/sgv2/docsapi/api/v2/namespaces/collections/documents/DocumentWriteResourceIntegrationTest.java:300]: same as above

**Review by ivansenic — APPROVED** [review]
Seems good, fix the comment and merge on green..

File: sgv2-docsapi/src/test/java/io/stargate/sgv2/docsapi/api/v2/namespaces/collections/documents/DocumentWriteResourceIntegrationTest.java [hunk 6]
```diff
@@ -128,7 +177,36 @@ public void withTtl() {
...
+              .body(DEFAULT_PAYLOAD)
+              .when()
+              .post(BASE_PATH, DEFAULT_NAMESPACE, DEFAULT_COLLECTION)
+              .peek();
+
+      assertThat(postResponse.statusCode()).isEqualTo(201);
+      assertThat(postResponse.header("location")).isNotNull();
+      String location = postResponse.header("location");
+
+      Awaitility.await()
```
Comment (ivansenic) [inline sgv2-docsapi/src/test/java/io/stargate/sgv2/docsapi/api/v2/namespaces/collections/documents/DocumentWriteResourceIntegrationTest.java:208]: here the idea was to assert document is there after writting before ttl expires

File: sgv2-docsapi/src/test/java/io/stargate/sgv2/docsapi/api/v2/namespaces/collections/documents/DocumentWriteResourceIntegrationTest.java [hunk 7]
```diff
@@ -152,6 +230,68 @@ public void malformedJson() {
...
+          .header("Content-Type", "application/json")
+          .body("[]")
+          .when()
+          .post(BASE_PATH, DEFAULT_NAMESPACE, DEFAULT_COLLECTION)
+          .then()
+          .body("code", equalTo(400))
+          .body(
+              "description",
+              equalTo(
+                  "Updating a key with just an empty object or an empty array is not allowed. Hint: update the parent path with a defined object instead."));
```
Comment (ivansenic) [inline sgv2-docsapi/src/test/java/io/stargate/sgv2/docsapi/api/v2/namespaces/collections/documents/DocumentWriteResourceIntegrationTest.java:263]: Same here?

**Review by ivansenic — CHANGES_REQUESTED** [review]
Something is wrong, why is writing the empty object or empty array not allowed here? Do we have the bug here in V1 as well?

OK seems that this is V1 behavior as well, and it was like this since forever apparently as `POST` was using the `documentService#putAtPath`.. However, this seems extremely off to me, let's sync please..

File: sgv2-docsapi/src/test/java/io/stargate/sgv2/docsapi/api/v2/namespaces/collections/documents/DocumentWriteResourceIntegrationTest.java [hunk 8]
```diff
@@ -152,6 +230,68 @@ public void malformedJson() {
...
+          .when()
+          .post(BASE_PATH, DEFAULT_NAMESPACE, DEFAULT_COLLECTION)
+          .then()
+          .statusCode(400)
+          .body("code", equalTo(400))
+          .body(
+              "description",
+              equalTo(
+                  "Updating a key with just an empty object or an empty array is not allowed. Hint: update the parent path with a defined object instead."));
+    }
```
[the remark this answers is not in the document]
Comment (EricBorczuk) [reply sgv2-docsapi/src/test/java/io/stargate/sgv2/docsapi/api/v2/namespaces/collections/documents/DocumentWriteResourceIntegrationTest.java:249]: Made a ticket for this (and empty array): https://datastax.jira.com/browse/C2-1899

**Pull request merged**